File: src/main/java/net/bindernews/eventbus/IPriority.java
package net.bindernews.eventbus;

import org.jetbrains.annotations.NotNull;

/**
 * Interface with default implementations allowing functional interfaces to easily add a "priority" feature.
 */
public interface IPriority<T> extends Comparable<T> {

    default int getPriority() {
        return 0;
    }

    @Override
    default int compareTo(@NotNull T o) {
        int d = getPriority() - ((IPriority<?>)o).getPriority();
        if (d != 0) {
            return d;
        } else {
            return hashCode() - o.hashCode();
        }
    }
}

File: src/main/kotlin/net/bindernews/grackle/interfaces/IPriority.kt
package net.bindernews.grackle.interfaces

/**
 * Interface with default implementations allowing functional interfaces to easily add a "priority" feature.
 */
interface IPriority : Comparable<IPriority> {
    val priority: Int get() = 0

    override fun compareTo(other: IPriority): Int {
        val d = priority - other.priority
        return if (d != 0) { d } else { hashCode() - other.hashCode() };
    }
}